src/vcu/src/controls.c

```diff
@@ -22,6 +29,123 @@ void init_controls_defaults(void) {
   control_settings.volt_lim_min_gain = 0;
   control_settings.volt_lim_min_voltage = 300;
   control_settings.torque_temp_limited = false;
+
+
+  power_limiting_settings.max_power = 80; //kW
+  power_limiting_settings.electrical_P = 1; 
+  power_limiting_settings.electrical_I = 1;
+  power_limiting_settings.anti_windup = 1; 
+  power_limiting_settings.pl_enable = false; 
+
+
+}
+/*
+int32_t get_electrical_power_limited_torque(int32_t pedal_torque) { 
+  if (cs_readings.power < power_limiting_settings.max_power * 100) {
+    accumulated_torque_error = 0;
+    return pedal_torque;
+
+
+  }
+  else {
+    int32_t power_error = (cs_readings.power - (power_limiting_settings.max_power * 100)); //This is now a positive value when over the limit
+    printf("POwer Error: %d \r \n", power_error);
+    int32_t current_speed = mc_readings.speed * 628 / 6000; //rad/s = rpm * 2pi/60 
+    int32_t torque_error = -1*power_error / current_speed * 10; //dNm, sign flipped due to speed direction
+    accumulated_torque_error = accumulated_torque_error + torque_error; 
+    printf("Accumulated error: %d \r \n", accumulated_torque_error);
+    if (accumulated_torque_error > power_limiting_settings.anti_windup) { 
+      accumulated_torque_error = power_limiting_settings.anti_windup; 
+    }
+
+    int32_t limited_torque = pedal_torque - (power_limiting_settings.electrical_P * torque_error + power_limiting_settings.electrical_I * accumulated_torque_error); 
+
+    if (limited_torque < 0) {
+      limited_torque = 0;
+    }
+
+    if (limited_torque < pedal_torque){ 
+      return limited_torque;
+    }
+    else { 
+      return pedal_torque; 
+    }
+
+    //check for big jumps in torque
+
+  }
+
+}
+*/
+
+int32_t get_electrical_power_limited_torque(int32_t pedal_torque) { 
+    int32_t power_error = (cs_readings.power - (power_limiting_settings.max_power * 100)); //This is now a positive value when over the limit, negative when under
+    int32_t limited_torque = 0;
+
+    printf("Power Error: %d \r \n", power_error);
+    int32_t current_speed = mc_readings.speed * 628 / 6000; //rad/s = rpm * 2pi/60 
+    int32_t torque_error = -1*power_error / current_speed * 10; //dNm, sign flipped due to speed direction
+    accumulated_torque_error = accumulated_torque_error + torque_error; 
+
+    if (accumulated_torque_error > power_limiting_settings.anti_windup) { 
+      accumulated_torque_error = power_limiting_settings.anti_windup; 
+    }
+    else if (accumulated_torque_error < -1 * power_limiting_settings.anti_windup) { 
+      accumulated_torque_error = -1*power_limiting_settings.anti_windup;
+    }
+    printf("Accumulated error: %d \r \n", accumulated_torque_error);
```

Comment:
Generally I would recommend against leaving in debugging print statements in stuff